### ISSUE TITLE
Boxed value types and String(s) should not be equal

### ISSUE THREAD

**Issue by michaelcheers** [bridgedotnet/Bridge#1357]
### Expected

False
### Actual

True
### Steps To Reproduce

Type k and k again.
[Live Bridge](http://live.bridge.net/#8f3d57ec2ef3fa522eb5aea4422eaa2a)
[Dot Net Fiddle](https://dotnetfiddle.net/qJYsWz)

``` csharp
public class Program
{
    static object TryRead ()
    {
        string readline = Console.ReadLine();
        int value;
        if (int.TryParse(readline, out value))
            return value;
        else
            return readline;
    }
    [Ready]
    public static void Main()
    {
        object a = TryRead();
        object b = TryRead();
        Console.WriteLine(a == b);
    }
}
```

**Comment by michaelcheers** [comment]
Also, if you type '7' then '7' again it has the same problem.

**Comment by LeonidVeriga** [comment]
Yes, @michaelcheers, thank you, we reproduced the problem.

I just simplified you example:
[Live Bridge](http://live.bridge.net/#d98afb975c5e834b6acbb1d3cd1ce120)
[Dot Net Fiddle](https://dotnetfiddle.net/V0f7HE)
```
public class Program
{
    [Ready]
    public static void Main()
    {
        object a = Convert.ToString(3);
        object b = Convert.ToString(3);

        var r = a == b;
        Console.WriteLine("Should be false for strings: " + r);

        object a1 = 7;
        object b1 = 7;

        var r1 = a1 == b1;
        Console.WriteLine("Should be false for int: " + r1);
    }
}
```

**Comment by weltkante** [comment]
> String (s) which are Object(s) should never be equal.

and

> Boxed value types and String(s) should never be equal.

This titles are misleading, they can be equal, just not in the sample code. It is easy to write code using boxed values or strings which are stored in object-variables and actually **must be** equal.

So the "never" in "should never be equal" is definitely wrong.

I adjusted the sample code to present an example where the comparison must be true.

```
using System;
					
public class Program
{
	public static void Main()
	{
        object a = Convert.ToString(3);
        object b = a;
		
		var r = a == b;
		
        Console.WriteLine("Should be true for strings: " + r);
		
		object a1 = 7;
        object b1 = a1;
		
		var r1 = a1 == b1;
		
        Console.WriteLine("Should be true for int: " + r1);
	}
}
```

**Comment by weltkante** [comment]
Also for strings, relying on them being different is (probably) relying on an implementation detail of the .NET runtime, due to string interning. Different .NET runtimes are allowed to behave differently in regards to which strings are automatically interned and which not, for example [some versions of .NET will automatically intern all copies of the empty string to the same one](https://blogs.msdn.microsoft.com/ericlippert/2009/09/28/string-interning-and-string-empty/).

Unless I missed something in the .NET specs that forbids it, runtimes would theoretically be allowed to intern all strings, making it behave like JS currently does. Of course native runtimes don't do that since it comes at a performance cost they don't want to pay.

So basically it is a choice on your side whether you want to replicate that implementation detail for strings in your JS-based runtime, in order to care for buggy programs which rely on that implementation detail. (Note that this is only for strings and not for boxed values, for boxed values the behavior is clearly specified, every time a value is boxed a new copy is created.)

**Comment by LeonidVeriga** [comment]
@weltkante, thank you for the comments.

Yes, the same boxed object equals to themselves. I corrected the title.

Regarding string interning, as far as I know, in JavaScript world it also depends on implementation details.

**Comment by LeonidVeriga** [comment]
When implementing, check this case as well - https://github.com/bridgedotnet/Bridge/issues/1514
